### PR TITLE
Add acc-provision role to support acc-provision operator

### PR DESCRIPTION
**Pull request by apoorva11029**
Relevant files:

watches.yaml: Tells the operator which CRD to watch for
requirements.yaml: External Ansible libraries required
Dockerfile: Change acc-provision branch name here according to release
roles/acc-provision: Business logic which defines what tasks run when CR updates happen

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+FROM quay.io/operator-framework/ansible-operator:v1.3.0
```
Comment (snaiksat) [inline Dockerfile:1]: Can we check if this is ubi-8 based?

File: roles/accprovision/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
+Role Name
```
Comment (snaiksat) [inline roles/accprovision/README.md:1]: Seems like this is auto-generated placeholder. Should we be putting some high level info here?

File: roles/accprovision/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+users too:
+
+    - hosts: servers
+      roles:
+         - { role: username.rolename, x: 42 }
+
+License
+-------
+
+BSD
```
Comment (snaiksat) [inline roles/accprovision/README.md:38]: Remove this section.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+COPY requirements.yml ${HOME}/requirements.yml
+USER 0
+RUN update-crypto-policies --set LEGACY && pip3 install pyopenssl
+RUN ansible-galaxy collection install -r ${HOME}/requirements.yml \
+ && chmod -R ug+rwx ${HOME}/.ansible
+RUN yum install git -y
+RUN git clone --single-branch --branch master https://github.com/noironetworks/acc-provision.git
```
Comment (snaiksat) [inline Dockerfile:9]: Could we make the branch an argument (with default being master)? That way we don't have to keep updating this file for every release. The build job will pass the branch name as a build arg.

File: Dockerfile (hunk 1 above)
Comment (apoorva11029) [reply Dockerfile:1]: This is a ubuntu20.04 image

Comment (snaiksat) [reply Dockerfile:1]: That does not work, our images all have to be ubi-8 based.

**Comment by snaiksat** [comment]
1. The image needs to be ubi-8 based.
2. The comment about the branch name being passed as an argument is marked resolved but I don't see that its been addressed.

File: Dockerfile (hunk 1 above)
Comment (apoorva11029) [reply Dockerfile:1]: My bad, it is ubi based 
https://github.com/operator-framework/operator-sdk/blob/f8f3690d9a3198d2f63411ea2ac53b2bce0c3c9d/images/ansible-operator/Dockerfile#L1
Checked on the running operator too

bash-4.4$ cat /etc/os-release 
NAME="Red Hat Enterprise Linux"
VERSION="8.3 (Ootpa)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="8.3"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Red Hat Enterprise Linux 8.3 (Ootpa)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8.3:GA"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 8"
REDHAT_BUGZILLA_PRODUCT_VERSION=8.3
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.3"

**Comment by apoorva11029** [comment]
> 1. The image needs to be ubi-8 based.
> 2. The comment about the branch name being passed as an argument is marked resolved but I don't see that its been addressed.

1. Confirmed about the ubi8 base and also posted more details here https://github.com/noironetworks/acc-provision-operator/pull/2#discussion_r677810963
2. Made the branch an argument called ACC_PROVISION_REPO_BRANCH that's set to master by default

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,18 @@
+FROM quay.io/operator-framework/ansible-operator:v1.3.0
+ARG ACC_PROVISION_REPO_BRANCH
+ENV ACC_PROVISION_BRANCH=${ACC_PROVISION_REPO_BRANCH:-master}
+
+COPY requirements.yml ${HOME}/requirements.yml
+USER 0
```
Comment (abhis2112) [inline Dockerfile:6]: do we require root user ?

File: molecule/default/tasks/accprovision_test.yml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    wait: yes
+    wait_timeout: 300
+    wait_condition:
+      type: Running
+      reason: Successful
+      status: "True"
+  vars:
+    cr_file: 'aci_v1alpha1_accprovision.yaml'
+
+- name: Add assertions here
```
Comment (abhis2112) [inline molecule/default/tasks/accprovision_test.yml:16]: is this task required?

File: roles/accprovision/defaults/main.yml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+---
+# defaults file for AccProvision
+acc_provision_dir_path: /tmp/acc_provision
+acc_provision_file_name: acc_provision_input.yaml
+acicnioperator_cr: cr.yaml
```
Comment (abhis2112) [inline roles/accprovision/defaults/main.yml:5]: Nomenclature cr.yaml might not be accurate

File: molecule/default/tasks/accprovision_test.yml [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    namespace: '{{ namespace }}'
+    definition: "{{ lookup('template', '/'.join([samples_dir, cr_file])) | from_yaml }}"
+    wait: yes
+    wait_timeout: 300
+    wait_condition:
+      type: Running
+      reason: Successful
+      status: "True"
+  vars:
+    cr_file: 'aci_v1alpha1_accprovision.yaml'
```
Comment (abhis2112) [inline molecule/default/tasks/accprovision_test.yml:14]: where we can find this file?

File: roles/accprovision/meta/main.yml [hunk 9]
```diff
@@ -0,0 +1,64 @@
+---
+galaxy_info:
+  author: your name
+  description: your description
```
Comment (abhis2112) [inline roles/accprovision/meta/main.yml:4]: do we need descriptions here?

File: roles/accprovision/scripts/process_string_to_yaml.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+def main():
+    cmap_input_path = os.getenv('CONFIGMAP_INPUT_FILE')
+    cmap = dict()
+    input_dict = dict()
+    with open(cmap_input_path, 'r') as stream:
+        cmap = yaml.safe_load(stream)
+        yaml_cmap_intermediate = cmap['data']['spec']
+        input_dict = yaml.safe_load(yaml_cmap_intermediate)
+        yaml_cmap = input_dict['acc_provision_input']
+    crd_input = os.getenv('CRDINPUT')
```
Comment (abhis2112) [inline roles/accprovision/scripts/process_string_to_yaml.py:17]: these 2 commands can be clubbed

File: roles/accprovision/scripts/process_string_to_yaml.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    yaml_crd = yaml.safe_load(crd_input)
+
+    crd_definition = os.getenv('CRD_DEFINITION')
+    crd_definition = yaml.safe_load(crd_definition)
```
Comment (abhis2112) [inline roles/accprovision/scripts/process_string_to_yaml.py:21]: likewise here i believe

File: roles/accprovision/scripts/process_string_to_yaml.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    cmap['metadata'].pop('creationTimestamp', None)
+    cmap['metadata'].pop('resourceVersion', None)
+    cmap['metadata'].pop('uid', None)
+    with open(cmap_input_path, 'w') as outfile:
+        json.dump(cmap, outfile)
+
+    # Write acc_provision_input to file
+    acc_prov_dir = os.getenv('ACCPROVDIR')
+    acc_prov_file = os.getenv('ACCPROVFILE')
+    acc_prov_file_path = os.path.join(acc_prov_dir, acc_prov_file)
```
Comment (abhis2112) [inline roles/accprovision/scripts/process_string_to_yaml.py:38]: can be clubbed above 2 vars

File: roles/accprovision/scripts/process_string_to_yaml.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    deep_merge(yaml_crd, yaml_cmap, crd_definition)
+
+    # Write configmap file
+    input_dict['acc_provision_input'] = yaml_crd
+    cmap['data']['spec'] = json.dumps(input_dict)
+    cmap['metadata'].pop('managedFields', None)
+    cmap['metadata'].pop('annotations', None)
+    cmap['metadata'].pop('creationTimestamp', None)
+    cmap['metadata'].pop('resourceVersion', None)
+    cmap['metadata'].pop('uid', None)
```
Comment (abhis2112) [inline roles/accprovision/scripts/process_string_to_yaml.py:31]: In metadata, there are other autogenerated fields as well to pop like "generation", selfLink . Check with the API schema supported in metadata for the crd definition. Do we need only the above things to be removed?

File: roles/accprovision/scripts/process_string_to_yaml.py (hunk 10 above)
Comment (abhis2112) [reply roles/accprovision/scripts/process_string_to_yaml.py:17]: yaml_crd = yaml.safe_load(os.getenv('CRDINPUT'))

File: roles/accprovision/tasks/process.yaml [hunk 14]
```diff
@@ -0,0 +1,80 @@
+---
+- name: Create temporary directory for acc-provision run
+  file:
+    path: "{{ acc_provision_dir_path }}"
+    state: directory
+    mode: '0777'
```
Comment (abhis2112) [inline roles/accprovision/tasks/process.yaml:6]: Do we need all access - Might be 644?

File: roles/accprovision/tasks/process.yaml [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+  script: "scripts/process_string_to_yaml.py"
+  args:
+    executable: /usr/bin/python3
+  environment:
+    CRDINPUT: "{{ acc_provision_input }}"
+    CRD_DEFINITION: "{{ crd_definition }}"
+    CONFIGMAP_INPUT_FILE: "{{ acc_provision_dir_path }}/cmap.yaml"
+    ACCPROVDIR: "{{ acc_provision_dir_path }}"
+    ACCPROVFILE: "{{ acc_provision_file_name }}"
+  register: hello
```
Comment (abhis2112) [inline roles/accprovision/tasks/process.yaml:50]: registered variable name

File: roles/accprovision/tasks/process.yaml [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+    namespace: "{{ lookup('env', 'WATCH_NAMESPACE') }}"
+    wait: yes
+
+- name: Create new configmap
+  k8s:
+    state: present
+    force: yes
+    src: "{{ acc_provision_dir_path }}/cmap.yaml"
+
+- name: Generate acicnioperator CR
```
Comment (abhis2112) [inline roles/accprovision/tasks/process.yaml:67]: Does this check if acicnioperator CR is already created or not. Will it overwrite?

**Review by abhis2112 — CHANGES_REQUESTED** [review]
Few comments

File: roles/accprovision/defaults/main.yml (hunk 7 above)
Comment (apoorva11029) [reply roles/accprovision/defaults/main.yml]: We have now moved to creating a deployment file instead and called the variable aci_cni_deployment_file

File: molecule/default/tasks/accprovision_test.yml (hunk 6 above)
Comment (apoorva11029) [reply molecule/default/tasks/accprovision_test.yml:16]: The molecule directory is a placeholder for future tests to be added

File: roles/accprovision/tasks/process.yaml (hunk 15 above)
Comment (apoorva11029) [reply roles/accprovision/tasks/process.yaml:50]: Removed, thanks for noticing

File: roles/accprovision/tasks/process.yaml (hunk 16 above)
Comment (apoorva11029) [reply roles/accprovision/tasks/process.yaml]: We have now switched to creating a deployment file- and the k8s module in Ansible patches the YAML objects. https://docs.ansible.com/ansible/latest/collections/community/kubernetes/k8s_module.html#parameter-state

File: roles/accprovision/scripts/process_string_to_yaml.py (hunk 12 above)
Comment (apoorva11029) [reply roles/accprovision/scripts/process_string_to_yaml.py:38]: We need the ACCPROVDIR variable in other places too, so made two separate variables

File: roles/accprovision/scripts/process_string_to_yaml.py (hunk 13 above)
Comment (apoorva11029) [reply roles/accprovision/scripts/process_string_to_yaml.py:31]: These are the following fields I see in the configmap schema, do we need anything else to be removed then?:
f:metadata:
        f:labels:
          .: {}
          f:aci-containers-config-version: {}
          f:network-plugin: {}
        f:ownerReferences:
          .: {}
          k:{"uid":"e3492916-4ce2-469e-80da-1f07c89a11f7"}:
            .: {}
            f:apiVersion: {}
            f:kind: {}
            f:name: {}
            f:uid: {}

File: Dockerfile (hunk 5 above)
Comment (apoorva11029) [reply Dockerfile:6]: Tried without it but you must be root to run update-crypto-policies.

File: roles/accprovision/tasks/process.yaml (hunk 14 above)
Comment (apoorva11029) [reply roles/accprovision/tasks/process.yaml]: We need to create and read files in that directory

Comment (abhis2112) [reply roles/accprovision/tasks/process.yaml]: 777 is for whole access. like running a script. For directory I believe 755 also works

File: roles/accprovision/scripts/process_string_to_yaml.py (hunk 13 above)
Comment (abhis2112) [reply roles/accprovision/scripts/process_string_to_yaml.py:31]: Check it via actual configmap running on the cluster. You will get the idea. The fields are added by apiserver which are autogenerated.

File: roles/accprovision/scripts/process_string_to_yaml.py (hunk 12 above)
Comment (abhis2112) [reply roles/accprovision/scripts/process_string_to_yaml.py:38]: i was referring to make it like acc_prov_file_path = os.path.join(os.getenv('ACCPROVDIR'), os.getenv('ACCPROVFILE'))

I don't see acc_prov_dir  and acc_prov_file variables used anywhere else

File: Dockerfile (hunk 5 above)
Comment (abhis2112) [reply Dockerfile:6]: ok sure.

File: roles/accprovision/scripts/process_string_to_yaml.py (hunk 12 above)
Comment (abhis2112) [reply roles/accprovision/scripts/process_string_to_yaml.py:38]: It's a minor comment. Just trying to decrease the line of codes and number of variables. :)

File: roles/accprovision/scripts/process_string_to_yaml.py (hunk 13 above)
Comment (abhis2112) [reply roles/accprovision/scripts/process_string_to_yaml.py:31]: The only extra field I saw here which we should add if we want to remove autogenerated for configmap is "selfLink" .

File: roles/accprovision/scripts/process_string_to_yaml.py (hunk 12 above)
Comment (apoorva11029) [reply roles/accprovision/scripts/process_string_to_yaml.py:38]: Ah I see what you mean, sure let me merge those. Thanks

File: roles/accprovision/tasks/process.yaml (hunk 14 above)
Comment (apoorva11029) [reply roles/accprovision/tasks/process.yaml]: Changed it, works!

File: roles/accprovision/scripts/process_string_to_yaml.py (hunk 13 above)
Comment (apoorva11029) [reply roles/accprovision/scripts/process_string_to_yaml.py:31]: Added selfLink too, it's deprecated and removed in k8s1.21 but the code wont fail regardless

**Review by abhis2112 — APPROVED** [review]

**Pull request merged**